src/mds/Server.cc

```diff
@@ -4117,6 +4117,24 @@ void Server::handle_client_getattr(MDRequestRef& mdr, bool is_lookup)
     } else if (ref->filelock.is_stable() ||
 	       ref->filelock.get_num_wrlocks() > 0 ||
 	       !ref->filelock.can_read(mdr->get_client())) {
+      /* Since we're taking advantage of an optimization here:
+       *
+       * We cannot suddenly, due to a changing condition, add this filelock as
+       * it can cause lock-order deadlocks. In this case, that condition is the
+       * lock state changes between request retries. If that happens, we need
+       * to check if we've acquired the other locks in this vector. If we have,
+       * then we need to drop those locks and retry.
+       */
+      if (mdr->is_rdlocked(&ref->linklock) ||
+          mdr->is_rdlocked(&ref->authlock) ||
+          mdr->is_rdlocked(&ref->xattrlock)) {
+        /* start over */
+        dout(20) << " dropping locks and restarting request because filelock state change" << dendl;
+	mds->locker->drop_locks(mdr.get());
+	mdr->drop_local_auth_pins();
+	mds->queue_waiter(new C_MDS_RetryRequest(mdcache, mdr));
+        return;
+      }
```

Comment:
@batrick 

I am thinking whether should we mandatory the lock orders for the `CInode` everywhere. Because when acquiring the locks it seems won't release the previous locks being already acquired and then goes to sleep ?

I found in the `mdcache->path_traverse()` it will obey the order:

1, filelock
2, netstlock
3, authlock

And then in the callers will do:

4, linklock
5, snaplock  

But some places like in `Server::handle_client_rename()` it will break it.

I just suspect will this happen in other places.

Reply:
Usually the locks are sorted prior to acquisition and the locks acquired should only rely on the "request" state. Can you point to a specific place you're thinking is doing this wrong?

In any case, I think we could add code to the locker to detect out-of-order locking. Would you like to add that? :)

Reply:
Certainly.

Please see https://github.com/ceph/ceph/blob/main/src/mds/Server.cc#L8931-L8966 for `oldin`:

```
    if (oldin) {
      // xlock oldin (for nlink--)
      lov.add_xlock(&oldin->linklock);
      lov.add_xlock(&oldin->snaplock);
      if (oldin->is_dir()) {
	ceph_assert(srci->is_dir());
	lov.add_rdlock(&oldin->filelock);   // to verify it's empty
```

Reply:
Is the concern that `oldin` will change? Because we've locked the dentry paths, I don't think this is possible?

Reply:
Not very sure. I didn't dig it more yet.

Reply:
> Usually the locks are sorted prior to acquisition and the locks acquired should only rely on the "request" state. Can you point to a specific place you're thinking is doing this wrong?
> 
> In any case, I think we could add code to the locker to detect out-of-order locking. Would you like to add that? :)

I raised one tracker for this, and will work on it soon.

Reply:
> I raised one tracker for this, and will work on it soon.

Could you share this tracker @lxbsz (can't look it up on redmine).

Reply:
> > I raised one tracker for this, and will work on it soon.
> 
> Could you share this tracker @lxbsz (can't look it up on redmine).

Sure, sorry I just forgot to share it.

https://tracker.ceph.com/issues/62123

Reply:
We can continue to discuss about lock-order checks in the tracker.

This change LGTM.